src/testing.rst

```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
+
+    test:
+      imports:
+        - package_name
+
+Note that that ``package_name`` is the name that is imported by python,
+not necessarily the name of the conda package (they are sometimes different).
+
+Testing for an import will catch the bulk of the packaging errors, generally
+including presence of dependencies. However, it does not assure that the
+package works correctly particularly if it works correctly with the particular
+versions of dependencies used. So it is good to run some other test of the code itself.
+
+Running Unit Tests
+..................
+
+The trick here is that there are multiple ways to run unit tests in Python,
+including nose, pytest, etc.
+
+Also, some packages install the tests with the package, and thus they can be
+run in place. while others keep the tests with the source code, and thus can
+not be run straight from an installed package.
+
+Test Requirements
+.................
+
+Sometimes there are packages required to run the tests that are not required
+to simply use the package. This is usually a test-running framework, such as
+nose or pytest. You can ensure that it is included by adding it to the test
+stanza:
+
+.. code-block:: yaml
+
+    test:
+      imports:
+        - package_name
+    ...
+      requires:
+        - pytest
+
+Built-in Tests
+..............
+
+Some packages have testing built-in. In this case, you can put a test command
+directly in the test stanza:
+
+.. code-block:: yaml
+
+    test:
+      ...
+      commands:
+         python -c "import package_name; package_name.tests.runall()"
+
+Alternatively, you can add a file called ``run_test.py`` in the recipe that
+will be run at test time. This allows an arbitrarily complicated test script.
+
+pytest Tests
+............
+
+If the tests are installed with the package, pytest can find and run them
+for you with the following command::
+
+    test:
+      requires:
+        - pytest
+      commands:
+        - py.test --pyargs package_name
+
+nose Tests
+...........
+
+Someone please add a description and example for running nose on an installed package.
+
+
+Command Line Utilities
+......................
+
+If a python package installs command line utilities, you probably want to test that
+they were properly installed::
+
+    test:
+      commands:
+        - util_1 --help
+
+If the utility actually has a test mode, great. Otherwise simply invoking
+``--help`` or ``--version`` or something will at least test that it is
+installed and can run.
+
+Tests Outside of The Package
+............................
+
+Note that conda-build runs the tests in an isolated environment after installing
+the package -- thus, at this point it does not have access to the original source
+tarball.  This is to ensure that the test environment is as close as possible to
+what an end-user will see.
+
+This makes it very hard to run tests that are not installed with the package.
+
+**NOTE** if anyone has good ideas as to how to do that, please put it here!
```

Comment:
Right, that is this case. The `source_files` section says grab this from the unpacked tarball and copy them (recursively) into the test environment. This way we can use these selected files and directories in the test environment; even though, they were never installed into the prefix. Here's a link to the [docs]( https://conda.io/docs/building/meta-yaml.html#source-files ) though it is a bit terse. The conda-build 2 release [blog entry]( https://www.continuum.io/blog/developer-blog/whats-old-and-new-conda-build ) is a bit more detailed. Does that help clarify it?